### PR TITLE
Give ui.browseableMessage the ability to show copy and close buttons after the message, and add buttons to some instances (2nd try)

### LINKED ISSUE #14641: Report "Press escape to close" on NVDA dialogs which the user needs to press escape to close
### Is your feature request related to a problem? Please describe.

There are several places where a user will press a keystroke which opens a dialog which may not be obvious to all users.  Examples are:

NVDA+f pressed twice opens a dialog with the formatting of the current caret location
NVDA+r runs OCR and presents the results in a buffer the user can navigate
NVDA+k pressed twice (from NVDA 2023.1) opens a dialog with the destination URL of the current link.

Where opening something like NVDA settings should (generally) be obvious that the user is in a dialog,, these examples may not be.

### Describe the solution you'd like

My suggestion is to append "press escape to close" to the title of these dialogs / the information which is read when the user activates the keystroke.

### Describe alternatives you've considered

As it is currently, I think most users work it out, although it may still be confusing at first.  I think the proposal wouldn't inconvenience anyone, as the information would be read last and could be skipped / ignored.

### Additional context
Suggested during a conversation with user Jared via email.

### PR DESCRIPTION
**Pull request by XLTechie**
### Link to issue number:

Fixes #14641 
Addresses #16995
Addresses #16996 

### Summary of the issue:

In #14641 @Qchristensen reported that some users are confused by browseableMessages, and their lack of definite closure mechanisms.
During the conversation, it was pointed out that in some cases, a user might also desire a copy button.
During work on the PR (originally #16369), it was requested that the copy button be given an accelerator key.

### Description of user facing changes

Added copy and close buttons to some browseableMessages, and the capability to add them to others.

### Description of development approach

* Thanks to @michaelDCurran's change in `ui.browseableMessage`--namely adding a `Scripting.Dictionary` to carry arbitrary query-string-equivalent  style parameters to the `mshtml` instance behind `browseableMessage`, it is now possible to pass in values for two new buttons, and various translatable messages, without any contortions.
* Tried different methods in the JS and CSS of `message.html`, and eventually settled on one which displayed the two buttons side-by-side, under a separator. If neither button's label is supplied, the div containing the HR and buttons remains hidden.
* The "Copy" button can be activated with `Ctrl+Shift+C`, and the "Close" button by `Escape`. `ctrl+shift+c` is indicated to the user via an accessibility label.
* Fixed up some docstring parameter listings to match modern format.
* Per the issue, modified the Report Link Destination message, to have both a copy button and a close button.
* Per the issue, added buttons to the Report Character Information message.
* Per request in PR comments, used a live region to present a "Copy complete" or "copy failed" message to the user. This message remains on screen for five seconds.
* The AI suggested covering for more failures in the initialization and configuration of MSHTML. I used its suggested exception checking, but added a private function to display an error message to the user when one of the components fail, so at least some kind of user notice can be given. I based it on the already extant private function for messaging the user if a `browseableMessage` is called on a secure screen.
* Both of the warning functions for browseableMessage unavailable situations, are now self-contained, with respect to `wx`.
* Added an immediate return if the MSHTML window is opened without providing the dialog arguments (i.e. the message and title, at least). That situation should never actually occur, but in the unlikely event that a failure passes the error checking we have in `browseableMessage`, this will return immediately, instead of stranding the user in a blank window with no obvious close mechanism.

### Testing strategy:

* Tested using the report link destination window, that the basic functionality worked as expected.
* Tested that copying using the copy button, correctly formatted the clipboard result. That is so far only possible when using `.innerHTML`.
* Tested that presentation with screen layout off, does not include extra blank lines (per #16995). This can apparently only be achieved, by passing the message to the div using `.innerHTML`.
* Tested that the component failure error dialog works correctly, by causing a component to fail.
* Tested that the Can't run on UAC, etc. screens warning works correctly after changes.
 
### Known issues with pull request:

* `Ctrl+Shift+C` is not an ideal key to activate the copy button, and it can not be changed by translators.
* The shortcut `Ctrl+Shift+C` is reported, no matter the state of "Report object shortcut keys" in "Object Presentation" settings. This is, AFAIK, unavoidable.
* In addition to the others, it was requested that the OCR results window provide a close button. This is more difficult, and I'd rather leave it to a separate PR in case it is not desired after all.
* Security of text only messages, is currently only provided by Python's `html.escape()`. I don't believe security is degraded by this PR, but it is not improved, either.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Added "Copy" and customizable "Close" buttons to message dialogs, improving usability.
  - Enhanced key handling for closing windows and copying text using keyboard shortcuts.
  - Introduced user notifications for copy operation success or failure, providing immediate feedback.

- **Bug Fixes**
  - Improved error handling and user notifications for messaging components, ensuring reliability.

- **Documentation**
  - Updated user documentation to reflect new interactive elements and usability improvements.

### CODE REVIEW

File: source/message.html
Comment (XLTechie) [reply source/message.html:18]: @coderabbitai please rewrite this suggestion to take into account that MSHTML, where this code will be used, does not support the `const` keyword.

Comment (XLTechie) [reply source/message.html:18]: @coderabbitai That's good, but please present it in a committable GitHub suggestion, as is always preferred in this project.

File: source/message.html
Comment (XLTechie) [reply source/message.html:39]: @coderabbitai will this suggested code work in MSHTML (Internet Explorer)? If so, please present the code in a suggestion without use of the `const` keyword, which is unsupported in MSHTML.

Comment (XLTechie) [reply source/message.html:44]: @coderabbitai Unfortunately, this suggestion is incompatible with MSHTML. It is necessary to use the older selection and copy code.

**Comment by XLTechie** [comment]
@CyrilleB79 Here is a new version of this PR, which should address most of your concerns.

**Comment by seanbudd** [comment]
@XLTechie - can you explain how #16995 was addressed in this PR? what changed in the dev approach?

**Comment by XLTechie** [comment]
@seanbudd I've added a detailed answer to the description.

**Comment by CyrilleB79** [comment]
All seems nice, either with or without screen layout.

Only one remaining point:
Using `ui.browseableMessage("Hello world", "Trial")`, there are no buttons as expected. Though, if I press `control+shift+c`, nothing is reported (no "Text copied" message), but the text is in the clipboard.
IMO the copy shortcut should be disabled when the button is not present.

Also another point but it's not worth blocking the PR for this. The shortcut `control+shift+c` is reported, no matter the state of "Report object shortcut keys" in "Object Presentation" settings. Though, it's probably not possible to do anything to honour this setting, and this detail is really not significant at all. Just mentioning here for awareness and maybe to add it in known issues section of the PR.

**Comment by XLTechie** [comment]
@CyrilleB79 Thank you for catching that oversight. An easy fix, which should be in the next build.
I will add a note to the PR about the inability to keep the accelerator from being mentioned.

**Comment by XLTechie** [comment]
@seanbudd Can you say why this is milestoned for 2025.1 instead of 2024.4? I don't care particularly, but I had intended to work on the NH3 integration after this was merged, and before 2025.1.
I can still work on it before that, but am curious.

**Comment by XLTechie** [comment]
@coderabbitai full review

**Comment by XLTechie** [comment]
It's weird that this passed lint check a few days ago, but failed today after no changes to lintable files.

File: source/ui.py
Comment (XLTechie) [reply source/ui.py:129]: @coderabbitai, you appear to have suggested that existing code from one function, be moved to an inappropriate place in a different function. I believe you are confused. You should report yourself to your developers for this.

**Comment by XLTechie** [comment]
@coderabbitai resolve

**Comment by seanbudd** [comment]
@XLTechie - we try not to introduce new/risky features in the weeks before the first beta, it was unlikely that this would be merged before the first beta

**Comment by XLTechie** [comment]
@seanbudd Updated this to master. Note that the "Changes" section was missing from the changelog, and I have had to add it in this.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**
